### PR TITLE
mention dart2dart

### PR DESCRIPTION
**Pull request by sethladd**

### CODE REVIEW

**Comment by floitschG** [comment]
LGTM.

**Comment by zoechi** [comment]
What does this mean? dart2dart is dead?

**Comment by sethladd** [comment]
dart2dart has been unsupported for some time. In order to optimize dart2js, we need to split out dart2dart from dart2js. dart2js is optimizing to be a compile to JavaScript tool.

Do you use dart2dart ? If so, for what reasons?

**Comment by zoechi** [comment]
I don't use dart2dart myself, I just try to stay up-to-date what options are available in case I need a solution. Some people mentioned that they use it for deployment of server side code.

**Comment by sethladd** [comment]
We want to learn what server-side deployment requirements people have. Hopefully, with the new `.packages` file, it'll be much easier to create a directory that you can "zip and go".

**Comment by zoechi** [comment]
It's quit confusing. I found a comment from Bob in dartbug from about July 2014,  that it's uncommon that `pub build` is run on `bin`. In a recent comment on a DEP he mentioned just that as the only proper way for server-side deployment. I can't remember seeing some comment about this shift of direction. If it works and it is supported it is fine,  but it's hard to follow when there is no information.

**Comment by sethladd** [comment]
I understand. Today, there is no blessed way to bundle up a server-side Dart app. Of course, developers are creating their own tools to do this. I personally would love to see someone build a grinder task that covers this.

* creates a directory
* copies all the files over
* creates a packages/ dir
* copies all the dependencies over
* writes out the .packages file
* optional: zips it up

**Comment by zoechi** [comment]
I have yet to check if `pub build bin` does just that (without zip).

**Comment by sethladd** [comment]
I would assume `pub build` is just for web apps. Hence my suggestion to write a stand-alone tool or grinder task.

**Comment by zoechi** [comment]
I tried it and it seems to work fine. 
See also https://code.google.com/p/dart/issues/detail?id=15731#c13
```
transformers:
  - $dart2js:
      $exclude: "bin"
```
is helpful to not generate JS for `bin/*`

**Comment by zoechi** [comment]
I have to correct myself - it doesn't work. The files put into `build/bin` are not really related to `bin`.

**Comment by zoechi** [comment]
I just found the discussion where Bob mentioned to use `pub build` for bin recently https://github.com/lrhn/dep-pkgspec/issues/13

**Comment by tosh** [comment]
I've been using `dart2dart` to create a single-file command line tool that I use within my vagrant dev setup but I'll find another way than `dart2dart` :)

related: I just posted to the dart-core thread on resource management (https://groups.google.com/a/dartlang.org/forum/#!topic/core-dev/8WrVpjD6tt8). I feel like I don't have a good grasp of what the responsibilities of `pub` are (package mgmt and building client-side apps? + running transformers?) and how it relates to other tools like `dart2dart`, `dart2js`, `dart`, `grind`, `build.dart` files and custom built stuff (it seems like everyone has their own way to build and deploy dart apps (frontend & backend)).

imho nothing wrong to have one tool for various related use cases. I think `cargo` in the rust world is used for package management and building stuff whereas ruby has things separated in gem/bundler and rake (?).

dependency management and building seem naturally related. I also like the idea of using one tool as interface for package management, running a dev server, building/bundling etc + having a nice command line ui that helps me out (the pub is a one stop shop for drinks, snacks, music and watching sports right?).

I'm glad that we already have great dependency management situation (esp compared to other ecosystems) but how to build and deploy dart apps could be simpler and easier.

**Comment by eernstg** [comment]
Hello Günter,

I saw that discussion, too, but I don't think that Bob's description
contradicts the "zip build" story: The first approach he mentions for
server-side deployment is

    'Run pub build and copy the build directory into your container.
Currently, that will include all your dependencies since packages in the
build output contains the actual content'.

Later he mentioned '$exclude'ing dart2js, but the 'container' is the
deployment host, possibly under AppEngine or similar, and 'copy' might as
well be 'zip/unzip'. Comparing to Seth's bullet list, the only missing item
is 'writes out the .packages file', but that is a new thing which may be
introduced when the package spec DEP is finalized.

But you seemed to imply that it does not work .. or what did you mean when
you said 'not really related to bin'? Are you using  pub build --mode=*X*
bin where *X* is anything other than 'release' in order to get the *.dart
files transferred to build?

On Wed, May 20, 2015 at 10:21 AM, Günter Zöchbauer <notifications@github.com
> wrote:

> I just found the discussion where Bob mentioned to use pub build for bin
> recently lrhn/dep-pkgspec#13
> <https://github.com/lrhn/dep-pkgspec/issues/13>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/dart-lang/sdk/pull/6#issuecomment-103805883>.
>



-- 
Erik Ernst  -  Google Danmark ApS
Skt Petri Passage 5, 2 sal, 1165 København K, Denmark
CVR no. 28866984

**Comment by zoechi** [comment]
I just run `pub build bin` which used `--mode=release` by default. I have excluded `bin` for the `$dart2js` transformer (otherwise I get a `build/bin/main.dart.js` file). The files from `bin` aren't copied to `build/bin` and the packages copied to `build/bin/packages` are just a subset of the packages in `packages/` and not related to the code in `bin/main.dart`. Therefore to me it looks like `pub build` doesn't exclude `bin` when it's run but doesn't do anything reasonable with it either.

**Comment by eernstg** [comment]
Sounds like you should try `pub build --mode=x bin` (anything, just not
--mode=release), such that the dart files are transferred, including the
ones that are needed in `build/packages`.

On Wed, May 20, 2015 at 12:02 PM, Günter Zöchbauer <notifications@github.com
> wrote:

> I just run pub build bin which used --mode=release by default. I have
> excluded bin for the $dart2js transformer (otherwise I get a
> build/bin/main.dart.js file). The files from bin aren't copied to
> build/bin and the packages copied to build/bin/packages are just a subset
> of the packages in packages/ and not related to the code in bin/main.dart.
> Therefore to me it looks like pub build doesn't exclude bin when it's run
> but doesn't do anything reasonable with it either.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/dart-lang/sdk/pull/6#issuecomment-103832546>.
>



-- 
Erik Ernst  -  Google Danmark ApS
Skt Petri Passage 5, 2 sal, 1165 København K, Denmark
CVR no. 28866984

**Comment by zoechi** [comment]
Right, with a different `mode` the output looks better. Thanks a lot for this hint! 
But this is still a bug. I'm not able to build client and server at once. When I build `web` or `bin`, the previous content of `build` is purged and I can't build them at once because `web` requires `--mode=release`.

**Comment by zoechi** [comment]
http://dartbug.com/23495

**Comment by eernstg** [comment]
But why isn't it more useful, anyway, to 'pub build --mode=bin bin' for the
server, then zipping that into ~/server.zip, then 'pub build' (that is 'pub
build --mode=release web') for the client, then zipping that into
~/client.zip, and then deploying only the files needed for each deployment
target?

On Wed, May 20, 2015 at 1:29 PM, Günter Zöchbauer <notifications@github.com>
wrote:

> http://dartbug.com/23495
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/dart-lang/sdk/pull/6#issuecomment-103849622>.
>



-- 
Erik Ernst  -  Google Danmark ApS
Skt Petri Passage 5, 2 sal, 1165 København K, Denmark
CVR no. 28866984

**Comment by sethladd** [comment]
Love the discussion, but please take it to cloud@dartlang.org (where server-side issues are discussed). Thanks!

**Comment by sethladd** [comment]
We moved the conversation over to https://groups.google.com/a/dartlang.org/forum/#!topic/cloud/M538bS3xVPI

Please join us there!
